File: android/core/src/main/kotlin/io/github/notkanaa/equipe/core/viewmodel/LoginViewModel.kt
package io.github.notkanaa.equipe.core.viewmodel

import io.github.notkanaa.equipe.core.AppError
import io.github.notkanaa.equipe.core.AppServices
import io.github.notkanaa.equipe.core.InputValidation
import kotlin.coroutines.cancellation.CancellationException

// Port of TeamTasksCore/ViewModels/Auth/LoginViewModel.swift.

/** State of the « Connexion » screen. */
data class LoginState(
    val email: String = "",
    val password: String = "",
    val isSubmitting: Boolean = false,
    override val error: ErrorState? = null,
) : ErrorHolder {
    /** Both fields filled and no request in progress. */
    val canSubmit: Boolean
        get() = InputValidation.trimmed(email).isNotEmpty() && password.isNotEmpty() && !isSubmitting
}

/**
 * « Connexion » screen. On success `AuthService.authStates()` emits the new session and `AppModel` switches to the
 * signed-in app: the screen has nothing else to do.
 */
class LoginViewModel(
    private val services: AppServices,
    email: String = "",
) : ScreenModel<LoginState>(LoginState(email = email), { state, error -> state.copy(error = error) }) {
    var email: String
        get() = current.email
        set(value) = update { it.copy(email = value) }

    var password: String
        get() = current.password
        set(value) = update { it.copy(password = value) }

    /** Signs in. Returns true on success (the password field is then cleared). */
    suspend fun signIn(): Boolean {
        if (current.isSubmitting) return false
        dismissError()
        if (current.password.isEmpty()) {
            present(MISSING_PASSWORD_MESSAGE)
            return false
        }
        val normalizedEmail = try {
            InputValidation.email(current.email)
        } catch (error: AppError) {
            present(error)
            return false
        }
        update { it.copy(isSubmitting = true) }
        try {
            services.auth.signIn(normalizedEmail, current.password)
            update { it.copy(password = "") }
            return true
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            present(error)
            return false
        } finally {
            update { it.copy(isSubmitting = false) }
        }
    }

    companion object {
        const val MISSING_PASSWORD_MESSAGE: String = "Saisissez votre mot de passe."
    }
}

File: Packages/TeamTasksKit/Sources/TeamTasksCore/ViewModels/Auth/LoginViewModel.swift
import Foundation
import Observation

/// « Connexion » screen. On success `AuthService.authStates()` emits the new session and `AppModel` switches to
/// the signed-in app: the view has nothing else to do.
@MainActor
@Observable
public final class LoginViewModel: ErrorPresenting {
    public var email: String
    public var password = ""
    public private(set) var isSubmitting = false
    public var error: ErrorState?

    private let services: AppServices

    public init(services: AppServices, email: String = "") {
        self.services = services
        self.email = email
    }

    /// Both fields filled and no request in progress.
    public var canSubmit: Bool {
        !InputValidation.trimmed(email).isEmpty && !password.isEmpty && !isSubmitting
    }

    /// Signs in. Returns true on success (the password field is then cleared).
    @discardableResult
    public func signIn() async -> Bool {
        guard !isSubmitting else { return false }
        error = nil
        guard !password.isEmpty else {
            present(message: "Saisis ton mot de passe.")
            return false
        }
        let normalizedEmail: String
        do {
            normalizedEmail = try InputValidation.email(email)
        } catch {
            present(error)
            return false
        }
        isSubmitting = true
        defer { isSubmitting = false }
        do {
            try await services.auth.signIn(email: normalizedEmail, password: password)
            password = ""
            return true
        } catch {
            present(error)
            return false
        }
    }
}
